planner/core/collect_column_stats_usage.go

```diff
@@ -19,27 +19,46 @@ import (
 	"github.com/pingcap/tidb/parser/model"
 )
 
-// predicateColumnCollector collects predicate columns from logical plan. Predicate columns are the columns whose statistics
-// are utilized when making query plans, which usually occur in where conditions, join conditions and so on.
-type predicateColumnCollector struct {
-	// colMap maps expression.Column.UniqueID to the table columns whose statistics are utilized to calculate statistics of the column.
-	colMap map[int64]map[model.TableColumnID]struct{}
+const (
+	collectPredicateColumns uint64 = 1 << iota
+	collectHistNeededColumns
+)
+
+// columnStatsUsageCollector collects predicate columns and/or histogram-needed columns from logical plan.
+// Predicate columns are the columns whose statistics are utilized when making query plans, which usually occur in where conditions, join conditions and so on.
+// Histogram-needed columns are the columns whose histograms are utilized when making query plans, which usually occur in the conditions pushed down to DataSource.
+// The set of histogram-needed columns is the subset of that of predicate columns.
+type columnStatsUsageCollector struct {
+	// collectMode indicates whether to collect predicate columns and/or histogram-needed columns
+	collectMode uint64
 	// predicateCols records predicate columns.
 	predicateCols map[model.TableColumnID]struct{}
+	// colMap maps expression.Column.UniqueID to the table columns whose statistics are utilized to calculate statistics of the column.
```

Comment:
Can you add more comments about the special cases to explain this? It's not easy to understand why an `expression.Column.UniqueID` needs more table columns to calculate.

Reply:
Done.